File: Sources/RecorderApp/Transcription/OpenAICompatibleProviderProfile.swift
import Foundation

struct OpenAICompatibleProviderProfile: Codable, Equatable, Sendable {
    static let currentSchemaVersion = 1
    static let hktBaseURLPrefix = "https://api.uat.bot-builder.pccw.com/v1/groups/"

    let schemaVersion: Int
    let providerKind: AIProviderKind
    let baseURL: URL
    let groupID: String?
    let asrModel: String
    let llmModel: String
    let language: String
    let prompt: String

    private enum CodingKeys: String, CodingKey {
        case schemaVersion, providerKind, baseURL, groupID, asrModel, llmModel, language, prompt
    }

    private init(schemaVersion: Int = currentSchemaVersion, providerKind: AIProviderKind, baseURL: URL, groupID: String?, asrModel: String, llmModel: String, language: String, prompt: String) {
        self.schemaVersion = schemaVersion
        self.providerKind = providerKind
        self.baseURL = baseURL
        self.groupID = groupID
        self.asrModel = asrModel
        self.llmModel = llmModel
        self.language = language
        self.prompt = prompt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        schemaVersion = try container.decode(Int.self, forKey: .schemaVersion)
        providerKind = try container.decodeIfPresent(AIProviderKind.self, forKey: .providerKind) ?? .openAICompatible
        baseURL = try container.decode(URL.self, forKey: .baseURL)
        groupID = try container.decodeIfPresent(String.self, forKey: .groupID)
        asrModel = try container.decode(String.self, forKey: .asrModel)
        llmModel = try container.decode(String.self, forKey: .llmModel)
        language = try container.decode(String.self, forKey: .language)
        prompt = try container.decode(String.self, forKey: .prompt)
    }

    static func validated(baseURLText: String, asrModel: String, llmModel: String, language: String, prompt: String) throws -> Self {
        let normalizedURL = try normalizedGenericURL(baseURLText)
        return try make(providerKind: .openAICompatible, baseURL: normalizedURL, groupID: nil, asrModel: asrModel, llmModel: llmModel, language: language, prompt: prompt)
    }

    static func hktValidated(groupID: String, asrModel: String, llmModel: String, language: String, prompt: String) throws -> Self {
        let normalizedGroupID = try validatedHKTGroupID(groupID)
        return try make(providerKind: .hktGenAI, baseURL: hktBaseURL(groupID: normalizedGroupID), groupID: normalizedGroupID, asrModel: asrModel, llmModel: llmModel, language: language, prompt: prompt)
    }

    static func hktBaseURL(groupID: String) -> URL {
        URL(string: hktBaseURLPrefix + groupID + "/openai")!
    }

    static func validatedHKTGroupID(_ value: String) throws -> String {
        // Deliberately do not trim: whitespace is not an ASCII digit and must be rejected.
        guard (1...32).contains(value.utf8.count), value.utf8.allSatisfy({ (48...57).contains($0) }) else {
            throw ProviderProfileValidationError.invalidHKTGroupID
        }
        return value
    }

    static func validatedPersisted(_ profile: Self) throws -> Self {
        guard profile.schemaVersion == currentSchemaVersion else {
            throw ProviderProfileValidationError.unsupportedSchemaVersion(profile.schemaVersion)
        }
        switch profile.providerKind {
        case .openAICompatible:
            guard profile.groupID == nil else { throw ProviderProfileValidationError.invalidProviderConfiguration }
            return try validated(baseURLText: profile.baseURL.absoluteString, asrModel: profile.asrModel, llmModel: profile.llmModel, language: profile.language, prompt: profile.prompt)
        case .hktGenAI:
            guard let groupID = profile.groupID else { throw ProviderProfileValidationError.invalidHKTGroupID }
            let validated = try hktValidated(groupID: groupID, asrModel: profile.asrModel, llmModel: profile.llmModel, language: profile.language, prompt: profile.prompt)
            guard validated.baseURL == profile.baseURL else { throw ProviderProfileValidationError.invalidProviderConfiguration }
            return validated
        }
    }

    private static func make(providerKind: AIProviderKind, baseURL: URL, groupID: String?, asrModel: String, llmModel: String, language: String, prompt: String) throws -> Self {
        let asr = asrModel.trimmingCharacters(in: .whitespacesAndNewlines)
        let llm = llmModel.trimmingCharacters(in: .whitespacesAndNewlines)
        let language = try validatedLanguage(language)
        guard !asr.isEmpty else { throw ProviderProfileValidationError.missingASRModel }
        guard !llm.isEmpty else { throw ProviderProfileValidationError.missingLLMModel }
        return Self(providerKind: providerKind, baseURL: baseURL, groupID: groupID, asrModel: asr, llmModel: llm, language: language, prompt: prompt.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    private static func validatedLanguage(_ value: String) throws -> String {
        let language = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard ["yue", "en", "zh"].contains(language) else {
            throw ProviderProfileValidationError.invalidLanguage
        }
        return language
    }

    private static func normalizedGenericURL(_ text: String) throws -> URL {
        let trimmedURL = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: trimmedURL), let scheme = components.scheme?.lowercased(), let host = components.host?.lowercased(), !host.isEmpty else { throw ProviderProfileValidationError.invalidBaseURL }
        guard components.user == nil, components.password == nil, components.query == nil, components.fragment == nil else { throw ProviderProfileValidationError.unsupportedURLComponents }
        guard scheme == "https" || (scheme == "http" && isLoopback(host)) else { throw ProviderProfileValidationError.insecureRemoteURL }
        var path = components.percentEncodedPath
        while path.count > 1 && path.hasSuffix("/") { path.removeLast() }
        if path.isEmpty || path == "/" { path = "/v1" } else if !path.hasSuffix("/v1") { path += "/v1" }
        components.scheme = scheme
        components.host = host
        components.percentEncodedPath = path
        guard let url = components.url else { throw ProviderProfileValidationError.invalidBaseURL }
        return url
    }

    private static func isLoopback(_ host: String) -> Bool {
        if host == "localhost" || host == "::1" || host == "[::1]" { return true }
        let octets = host.split(separator: ".", omittingEmptySubsequences: false)
        guard octets.count == 4, octets.first == "127", octets.allSatisfy({ guard let value = UInt8($0) else { return false }; return String(value) == $0 || $0 == "0" }) else { return false }
        return true
    }
}

enum ProviderProfileValidationError: LocalizedError, Equatable {
    case invalidBaseURL, unsupportedURLComponents, insecureRemoteURL, missingASRModel, missingLLMModel, invalidLanguage, invalidHKTGroupID, invalidProviderConfiguration, unsupportedSchemaVersion(Int)
    var errorDescription: String? {
        switch self {
        case .invalidBaseURL: "Enter a valid API base URL."
        case .unsupportedURLComponents: "The API URL cannot contain credentials, a query, or a fragment."
        case .insecureRemoteURL: "Remote providers must use HTTPS."
        case .missingASRModel: "Enter an ASR model identifier."
        case .missingLLMModel: "Enter an LLM model identifier."
        case .invalidLanguage: "Choose Cantonese, English, or Chinese."
        case .invalidHKTGroupID: "Enter a group ID containing 1 to 32 ASCII digits."
        case .invalidProviderConfiguration: "The saved provider configuration is invalid."
        case let .unsupportedSchemaVersion(version): "Provider profile version \(version) is not supported."
        }
    }
}

File: windows/src/Recorder.Application/AI/OpenAICompatibleProviderProfile.cs
using System.Text.Json.Serialization;

namespace TeamsRecorder.Windows.Application.AI;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum AIProviderKind { OpenAICompatible, HktGenAI }

/// <summary>
/// The versioned, non-secret part of an OpenAI-compatible provider setting.
/// API credentials deliberately live in a separate per-user DPAPI store.
/// </summary>
public sealed record OpenAICompatibleProviderProfile
{
    public const int CurrentSchemaVersion = 3;
    public const string HktBaseUrlPrefix = "https://api.uat.bot-builder.pccw.com/v1/groups/";
    public const string DefaultBaseUrl = "https://api.openai.com/v1";
    public const string DefaultAsrModel = "gpt-4o-transcribe";
    public const string DefaultLlmModel = "gpt-5.6-terra";

    [JsonPropertyName("schemaVersion")] public int SchemaVersion { get; init; }
    [JsonPropertyName("providerKind")] public AIProviderKind ProviderKind { get; init; } = AIProviderKind.OpenAICompatible;
    [JsonPropertyName("baseURL")] public string BaseUrl { get; init; } = string.Empty;
    [JsonPropertyName("groupID")] public string? GroupId { get; init; }
    [JsonPropertyName("asrModel")] public string AsrModel { get; init; } = string.Empty;
    [JsonPropertyName("llmModel")] public string LlmModel { get; init; } = string.Empty;
    [JsonPropertyName("language")] public string Language { get; init; } = string.Empty;
    [JsonPropertyName("prompt")] public string Prompt { get; init; } = string.Empty;
    [JsonPropertyName("meetingIntelligencePrompt")] public string MeetingIntelligencePrompt { get; init; } = string.Empty;

    public static OpenAICompatibleProviderProfile Default => Validated(
        DefaultBaseUrl, DefaultAsrModel, DefaultLlmModel, language: string.Empty, prompt: string.Empty);

    public static OpenAICompatibleProviderProfile Validated(
        string baseUrlText, string asrModel, string llmModel, string? language, string? prompt,
        string? meetingIntelligencePrompt = null)
    {
        var normalizedUrl = NormalizeBaseUrl(baseUrlText);
        var normalizedAsr = TrimRequired(asrModel, ProviderProfileValidationError.MissingAsrModel);
        var normalizedLlm = TrimRequired(llmModel, ProviderProfileValidationError.MissingLlmModel);
        return new()
        {
            SchemaVersion = CurrentSchemaVersion,
            ProviderKind = AIProviderKind.OpenAICompatible,
            BaseUrl = normalizedUrl,
            GroupId = null,
            AsrModel = normalizedAsr,
            LlmModel = normalizedLlm,
            Language = language?.Trim() ?? string.Empty,
            Prompt = prompt?.Trim() ?? string.Empty,
            MeetingIntelligencePrompt = ValidateMeetingIntelligencePrompt(meetingIntelligencePrompt)
        };
    }

    public static OpenAICompatibleProviderProfile ValidateStored(OpenAICompatibleProviderProfile value)
    {
        ArgumentNullException.ThrowIfNull(value);
        if (value.SchemaVersion is not (1 or 2 or CurrentSchemaVersion))
            throw new ProviderProfileException(ProviderProfileValidationError.UnsupportedSchemaVersion);
        var meetingPrompt = value.SchemaVersion == 1 ? string.Empty : value.MeetingIntelligencePrompt;
        if (value.SchemaVersion < 3 || value.ProviderKind == AIProviderKind.OpenAICompatible)
            return Validated(value.BaseUrl, value.AsrModel, value.LlmModel, value.Language, value.Prompt, meetingPrompt);
        return HktValidated(value.GroupId ?? string.Empty, value.AsrModel, value.LlmModel,
            value.Language, value.Prompt, meetingPrompt);
    }

    public static OpenAICompatibleProviderProfile HktValidated(
        string groupId,
        string asrModel,
        string llmModel,
        string? language,
        string? prompt,
        string? meetingIntelligencePrompt = null)
    {
        var normalizedGroupId = groupId.Trim();
        if (normalizedGroupId.Length is < 1 or > 32 || normalizedGroupId.Any(character => character is < '0' or > '9'))
            throw new ProviderProfileException(ProviderProfileValidationError.InvalidHktGroupId);
        return new()
        {
            SchemaVersion = CurrentSchemaVersion,
            ProviderKind = AIProviderKind.HktGenAI,
            BaseUrl = HktBaseUrlPrefix + normalizedGroupId + "/openai",
            GroupId = normalizedGroupId,
            AsrModel = TrimRequired(asrModel, ProviderProfileValidationError.MissingAsrModel),
            LlmModel = TrimRequired(llmModel, ProviderProfileValidationError.MissingLlmModel),
            Language = language?.Trim() ?? string.Empty,
            Prompt = prompt?.Trim() ?? string.Empty,
            MeetingIntelligencePrompt = ValidateMeetingIntelligencePrompt(meetingIntelligencePrompt),
        };
    }

    private static string ValidateMeetingIntelligencePrompt(string? value)
    {
        var prompt = value?.Trim() ?? string.Empty;
        if (prompt.Length > 16_384 || prompt.Any(character => character == '\0'))
            throw new ProviderProfileException(ProviderProfileValidationError.InvalidMeetingIntelligencePrompt);
        return prompt;
    }

    private static string TrimRequired(string? value, ProviderProfileValidationError error)
    {
        var trimmed = value?.Trim();
        if (string.IsNullOrWhiteSpace(trimmed)) throw new ProviderProfileException(error);
        return trimmed;
    }

    private static string NormalizeBaseUrl(string? value)
    {
        if (!Uri.TryCreate(value?.Trim(), UriKind.Absolute, out var uri) || string.IsNullOrWhiteSpace(uri.Host))
            throw new ProviderProfileException(ProviderProfileValidationError.InvalidBaseUrl);
        if (!string.IsNullOrEmpty(uri.UserInfo) || !string.IsNullOrEmpty(uri.Query) || !string.IsNullOrEmpty(uri.Fragment))
            throw new ProviderProfileException(ProviderProfileValidationError.UnsupportedUrlComponents);
        if (!uri.Scheme.Equals(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase) &&
            !(uri.Scheme.Equals(Uri.UriSchemeHttp, StringComparison.OrdinalIgnoreCase) && IsLoopback(uri.Host)))
            throw new ProviderProfileException(ProviderProfileValidationError.InsecureRemoteUrl);

        var path = uri.AbsolutePath.TrimEnd('/');
        if (string.IsNullOrEmpty(path)) path = "/v1";
        else if (!path.EndsWith("/v1", StringComparison.OrdinalIgnoreCase)) path += "/v1";
        var builder = new UriBuilder(uri) { Scheme = uri.Scheme.ToLowerInvariant(), Host = uri.Host.ToLowerInvariant(), Path = path, Query = string.Empty, Fragment = string.Empty };
        return builder.Uri.AbsoluteUri.TrimEnd('/');
    }

    private static bool IsLoopback(string host) =>
        host.Equals("localhost", StringComparison.OrdinalIgnoreCase) || host == "::1" ||
        (System.Net.IPAddress.TryParse(host, out var address) && System.Net.IPAddress.IsLoopback(address));
}

public enum ProviderProfileValidationError
{
    InvalidBaseUrl,
    UnsupportedUrlComponents,
    InsecureRemoteUrl,
    MissingAsrModel,
    MissingLlmModel,
    InvalidMeetingIntelligencePrompt,
    InvalidHktGroupId,
    UnsupportedSchemaVersion
}

/// <summary>Safe-to-display validation failure; it intentionally contains no submitted URL or credential.</summary>
public sealed class ProviderProfileException(ProviderProfileValidationError reason) : InvalidOperationException(BuildMessage(reason))
{
    public ProviderProfileValidationError Reason { get; } = reason;
    private static string BuildMessage(ProviderProfileValidationError reason) => reason switch
    {
        ProviderProfileValidationError.InvalidBaseUrl => "Enter a valid API base URL.",
        ProviderProfileValidationError.UnsupportedUrlComponents => "The API URL cannot contain credentials, a query, or a fragment.",
        ProviderProfileValidationError.InsecureRemoteUrl => "Remote providers must use HTTPS.",
        ProviderProfileValidationError.MissingAsrModel => "Enter an ASR model identifier.",
        ProviderProfileValidationError.MissingLlmModel => "Enter an LLM model identifier.",
        ProviderProfileValidationError.InvalidMeetingIntelligencePrompt => "The meeting intelligence prompt is too large or invalid.",
        ProviderProfileValidationError.InvalidHktGroupId => "Enter an HKT group ID containing 1 to 32 digits.",
        _ => "This provider profile version is not supported."
    };
}
